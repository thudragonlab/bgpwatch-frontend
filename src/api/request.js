import axios from 'axios'
import {
  Message,
  MessageBox
} from 'element-ui'
// import { getToken } from '@/utils/auth'
// import { API_BASE_URL } from '@/utils/url_config'

// create an axios instance
let oldCrrentView = localStorage.getItem('currentView')
const service = axios.create({
  // baseURL: '', // url = base url + request url
  // withCredentials: true, // send cookies when cross-domain requests
  timeout: 30000 * 2
  // * 20, // request timeout
})

// request interceptor
service.interceptors.request.use(
  config => {
    // do something before request is sent
    // console.log(config)
    if (config.useToken) {
      const userInfo = localStorage.getItem('userInfo')
      if (userInfo) {
        const token = JSON.parse(userInfo).token
        // console.log(JSON.parse(userInfo).token)
        // config.headers['content-type'] = 'application/x-www-form-urlencoded;charset=UTF-8';
        config.headers['Authorization'] = 'Bearer ' + token
      }
    }

    return config
  },
  // error => {
  //   // do something with request error
  //   console.log(error) // for debug
  //   return Promise.reject(error)
  // }
)

// response interceptor
service.interceptors.response.use(
  /**
   * If you want to get http information such as headers or status
   * Please return  response => response
   */

  /**
   * Determine the request status by custom code
   * Here is just an example
   * You can also judge the status by HTTP Status Code
   */
  response => {
    const {
      data,
      headers
    } = response
    if (headers['content-type'] === "text/csv; charset=utf-8") {
      const fileName = headers['content-disposition'].replace(/\w+; filename=(.*)/, '$1')
      console.log(fileName)
      // 此处当返回json文件时需要先对data进行JSON.stringify处理，其他类型文件不用做处理
    } else {
      const data = response.data
      if(data.lastUpdateTimeStamp && data.lastUpdateTimeStamp > 0){
        const lastUpdateTimeStamp = data.lastUpdateTimeStamp
        const nowCrrentView = localStorage.getItem('currentView')
        if(oldCrrentView !== nowCrrentView){
          localStorage.setItem('lastUpdateTimestamp',lastUpdateTimeStamp)
          oldCrrentView = nowCrrentView
        }else{
          const oldLastUpdateTimestamp = localStorage.getItem('lastUpdateTimestamp')
          if(!oldLastUpdateTimestamp){
            localStorage.setItem('lastUpdateTimestamp',lastUpdateTimeStamp)  
          }else{
            const newUpdateTimestamp = Math.max(Number(lastUpdateTimeStamp),Number(oldLastUpdateTimestamp))
            localStorage.setItem('lastUpdateTimestamp',newUpdateTimestamp)
          }
        }
        
      }
      
      if (!data.status) {
        MessageBox({
          message: data.statusCode || 'Error',
          type: 'error',
          duration: 5 * 1000
        })
        return Promise.reject(data)
      }
    }
    return data.data
   
  },
  error => {
    console.error(error.message) // for debug
    if (error.message && error.message.includes('aborted')) {
      return Promise.reject(error)
    }

    MessageBox({
      message: 'No data found.',
      type: 'error',
      duration: 1 * 1000
    })
    return Promise.reject(error)
  }
)

export default service