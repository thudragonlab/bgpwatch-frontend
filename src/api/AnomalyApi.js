import request from './request'
const MY_URL = '/myapi'
export function getEventlistByDate(st, et) {
    return request({
        methods: 'GET',
        url: `${MY_URL}/events?st=${st}&et=${et}&event_type=ALL`,
        timeout: 3 * 60 * 1000,
    })
}
export function getMoasEventlistByDate(st, et) {
    return request({
        methods: 'GET',
        url: `${MY_URL}/events?st=${st}&et=${et}&event_type=ALL_MOAS`,
        timeout: 3 * 60 * 1000,
    })
}


export function getEventDetail(detail_url) {
    // return new Promise((resolve,reject) => {
    //   return resolve(eventsDetail)
    // })
    return request({
        methods: 'GET',
        url: `${MY_URL}${detail_url}`
    })
}


