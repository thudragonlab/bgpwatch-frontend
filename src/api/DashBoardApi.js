import request from './request'
import axios from 'axios'
// const BASE_URL = 'http://10.110.62.34:5000/'
const BASE_URL = '/dbapi'
const BASE_URL2 = '/newapi'
const MY_URL = '/myapi'
const WZQ_API2 = '/wzqApi2'


export function getAsInfoByASN(asn) {
  return request({
    method: 'GET',
    url: `${BASE_URL}/api/dashboard/daily/asinfo?asn=${asn}`,

  })
}

export function getAsInfoByASN_myAPI(asn) {
  return request({
    method: 'GET',
    url: `${MY_URL}/asInfo?asn=${asn}`,

  })
}


export function getPeerPrefix(asn, {
  type,
  detailAsn
}) {
  return request({
    method: 'GET',
    url: `${BASE_URL}/api/dashboard/daily/peerprefix?asn=${asn}&${type}=${detailAsn}`,
  })
}



export function getAsRoute(asn, type) {
  return request({
    method: 'GET',
    url: `${MY_URL}/getAsRoute?asn=${asn}&type=${type}`,
  })
}

export function getPeerTypeDistribute(asn) {
  return request({
    method: 'GET',
    url: `${MY_URL}/getPeerTypeDistribute?asn=${asn}`,
  })
}


export function getAsInfoByList(asnList) {
  return request({
    method: 'POST',
    url: `${MY_URL}/getAsInfoByList`,
    data: asnList.map(as => parseInt(as))
  })
}

export function getRealTimeData() {
  return request({
    method: 'GET',
    url: `${MY_URL}/getRealTimeData`,
  })
}

export function getRealTimeDataInAWhile() {
  return request({
    method: 'GET',
    url: `${MY_URL}/getRealTimeDataInAWhile`,
  })
}

export function getASList(asn) {
  return request({
    method: 'GET',
    url: `${MY_URL}/getAsList?asn=${asn}`,
  })
}

// import test from '@/../public/static/temporaryData/1.json'
export function getAsPathByPrefix(prefix, asn,type) {
  // return new Promise((resolve,reject) => {
  //   resolve(test)
  // })
  return request({
    method: 'GET',
    url: `${MY_URL}/getAsPathByPrefix?prefix=${prefix}&asn=${asn}&type=${type}`,
  })
}


export function getRouteHistoryStats(asn, qtype = 'g') {
  return request({
    method: 'GET',
    url: `${MY_URL}/route_history_stats?asns=${asn.join(',')}&qtype=${qtype}`,
  })
}


export function getRoutingPath(asn, pfx, qtype = "from") {
  return request({
    method: 'GET',
    url: `${WZQ_API2}/api/routing_path?asn=${asn}&pfx=${pfx}&qtype=${qtype}`
  })
}


export function getPrefixList(asn, qstr = undefined) {
  return request({
    method: 'GET',
    url: `${MY_URL}/prefix_list?asn=${asn}${qstr ? `&qstr=${qstr}` :  ''}`
  })
}

export function getPrefixTotalByAsn(asn) {
  return request({
    method: 'GET',
    url: `${MY_URL}/getPrefixTotalByAsn?asn=${asn}`,
  })
}

export function getPeerList(asn,ipVersion) {
  return request({
    method: 'GET',
    url: `${MY_URL}/peer_list?asn=${asn}&ipversion=${ipVersion}`
  })
}


// http://10.110.62.37:5050/api/routing_path?asn=4538&pfx=59.66.196.0%2F24&qtype=to

export function getWhoisInfoByASN(asn){
  return request({
    method: 'GET',
    url: `${MY_URL}/get_whois_info_by_asn?asn=${asn}`
  })
}