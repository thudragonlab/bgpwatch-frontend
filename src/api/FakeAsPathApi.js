import request from './request'
const MY_URL = '/myapi'

export function getFakeAsPath(st, et) {
    return request({
        methods: 'GET',
        url: `${MY_URL}/getFakeAsPath?st=${st}&et=${et}`,
        timeout: 3 * 60 * 1000,
    })
}


export function getFakeAsPathRecordById(id){

    return request({
        methods: 'GET',
        url: `${MY_URL}/getFakeAsPathRecordById?id=${id}`,
        timeout: 3 * 60 * 1000,
    })

}