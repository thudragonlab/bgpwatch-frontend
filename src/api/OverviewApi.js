import request from './request'
import axios from 'axios'
// const BASE_URL = 'http://10.110.62.34:5000/'
const MY_URL = '/myapi'


export function getOverviewData(startTime, endTime,selectedOption,asn) {
    return request({
      methods: 'GET',
      url: `${MY_URL}/overview?start=${startTime}&end=${endTime}&cc=${selectedOption}&asn=${asn}`,
    })
  }


export function downloadCSV(startTime, endTime) {
    return request({
        methods: 'GET',
        url: `${MY_URL}/exportCSV?start=${startTime}&end=${endTime}`,
        responseType:'blob'
      })
}