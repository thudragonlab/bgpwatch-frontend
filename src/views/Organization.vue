<template>
  <div class="org-container">
    <el-row class="search-bar">
      <el-col :col="24">
        <el-input
          id="searchInput"
          :placeholder="`Please input ${placeholderMap[selectedType]}`"
          @input="inputContent"
          v-model="inputText"
        >
          <el-select v-model="selectedType" id="searchSelect" slot="prepend" style="width: 150px" @change="changeModel">
            <el-option v-for="item in options" :key="item.value" :label="item.label" :value="item.value"> </el-option>
          </el-select>
          <el-button slot="append" icon="el-icon-search" @click="doSearch"></el-button>
        </el-input>
      </el-col>
    </el-row>

    <div class="as-table-box">
      <div class="as-table">
        <pagination-table hiddenIndex :originData="tabldata" :showRule="showRule" loadDataFadeIn></pagination-table>
      </div>
    </div>
    <div class="update-time">LastUpdate:{{ lastUpdateTimestamp }}</div>
  </div>
</template>

<script>
import { getAsInfo } from '@/api/routeMonitor'
import ToDashboardColumn from '@/components/Column/ToDashboardColumn.vue'
export default {
  name: 'Organization',
  created() {},
  data() {
    return {
      tabldata: [],
      inputText: 'CERNET',
      lastUpdateTimestamp: '',
      options: [
        {
          value: 'searchASN',
          label: 'ASN',
        },
        {
          value: 'searchAsName',
          label: 'AS Name',
        },
        {
          value: 'searchOrg',
          label: 'Org Name',
        },
      ],
      showRule: [
        {
          prop: 'asn',
          label: 'asn',
          subComponent: ToDashboardColumn,
          rowFormat: row => {
            return row.asn
          },
        },

        {
          prop: 'as_name',
          label: 'AS Name',
        },
        {
          prop: 'org_name',
          label: 'Organization',
        },
      ],
      selectedType: 'searchAsName',
      searchASN: '',
      searchAsName: 'CERNET',
      searchOrg: '',
      propList: ['searchASN', 'searchAsName', 'searchOrg'],
      placeholderMap: {
        searchASN: 'asn',
        searchAsName: 'as name',
        searchOrg: 'as organization',
      },
    }
  },
  async mounted() {
    this.doSearch()
  },
  methods: {
    async doSearch() {
      const result = await getAsInfo(this.searchASN, this.searchAsName, this.searchOrg)
      this.lastUpdateTimestamp = new Date(parseInt(localStorage.getItem('lastUpdateTimestamp') * 1000))
        .toJSON()
        .split('T')[0]
      this.tabldata = result
    },
    changeModel(v) {
      this.inputText = ''
      this.propList.forEach(prop => {
        if (v !== prop) {
          this[prop] = ''
        }
      })
    },
    inputContent() {
      this[this.selectedType] = this.inputText
    },
  },
}
</script>

<style lang="scss" scoped>
.org-container {
  padding: 22px;
  background-color: #f0f0f0;
  height: 100%;
}

.as-table-box {
  border-radius: 20px;
  padding: 10px;
  background-color: #fff;
}

.as-table {
  // border-radius: 20px;
}

.search-bar {
  background-color: #fff;
  border-radius: 20px;
  padding: 10px;
}

.update-time {
  background-color: #fff;
  text-align: right;
  font-weight: bold;
  margin-top: 10px;
  padding-right: 10px;
  border-radius: var(--border-radius-deg);
}
</style>
