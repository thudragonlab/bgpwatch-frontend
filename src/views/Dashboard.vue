<template>
  <div class="dashboard-container fade-in">
    <div :class="['sticky-div', showSticktyRow ? 'show-search-bar' : 'hidden-search-bar']">
      <el-input
        class="search-input"
        v-model="inputText"
        :placeholder="`You can search by AS number, AS name, organization name`"
      >
        <!-- <template slot="prepend">ASN</template> -->
        <el-button slot="append" icon="el-icon-search" @click="chickSearch"></el-button>
      </el-input>
    </div>

    <el-row class="row-cc" id="search-row">
      <el-col :span="9" v-if="DashboardActiveName !== 'Routing Path'" class="show-input" style="text-align: left">
        <el-input
          class="search-input"
          v-model="inputText"
          :placeholder="`You can search by AS number, AS name, organization name`"
        >
          <el-button slot="append" icon="el-icon-search" @click="chickSearch"></el-button>
        </el-input>
      </el-col>
      <el-col :span="15" class="last-update">
        <div style="text-align: left">You can search by AS number, AS name, or organization name.</div>
        <div>Last Update:{{ lastUpdateTimestamp }}</div>
      </el-col>
    </el-row>
    <el-row v-if="inputAsn === ''">
      <pagination-table
        v-loading="showLoading"
        hiddenIndex
        :originData="tabldata"
        :showRule="showRule"
        loadDataFadeIn
        @clickAsn="clickAsn"
      ></pagination-table>
    </el-row>
    <el-row v-show="inputAsn !== ''">
      <el-tabs stretch v-model="DashboardActiveName" type="border-card" class="tab-list">
        <el-tab-pane
          v-for="(tab, index) in Object.keys(DashboardTabs)"
          :lazy="DashboardTabs[tab].lazy"
          :key="index"
          :label="DashboardTabs[tab].label"
          :name="DashboardTabs[tab].label"
        >
          <component :is="DashboardTabs[tab].subComponent" :param="DashboardTabs[tab].param"></component>
        </el-tab-pane>
      </el-tabs>
    </el-row>
  </div>
</template>

<script>
import AsInfoTab from '@/components/Dashboard/DashboardAsInfoTab.vue'
import DashboardWhoisTab from '@/components/Dashboard/DashboardWhoisTab.vue'
import IpTab from '@/components/Dashboard/DashboardIPTab.vue'

import { getAsInfo } from '@/api/routeMonitor'
import { setDataInGraphData } from '@/templates/chart/graphTemplate'
import { getPeerTypeDistribute, getAsInfoByList } from '@/api/DashBoardApi'
import ToDashboardColumn from '@/components/Column/ToDashboardColumn.vue'
export default {
  name: 'Dashboard',
  data() {
    return {
      selectedType: 'inputAsn',
      DashboardActiveName: 'Basic',
      inputText: localStorage.getItem('DashboardASN') || '4538',
      showLoading: true,
      inputAsn: localStorage.getItem('DashboardASN') || '4538',
      searchAsName: '',
      searchOrg: '',
      tabldata: [],
      lastUpdateTimestamp: '',
      propList: ['inputAsn', 'searchAsName', 'searchOrg'],
      showRule: [
        {
          prop: 'asn',
          label: 'asn',
          subComponent: ToDashboardColumn,
          rowFormat: row => {
            return row.asn
          },
        },

        {
          prop: 'as_name',
          label: 'AS Name',
        },
        {
          prop: 'org_name',
          label: 'Organization',
        },
        {
          prop: 'cone',
          label: 'Cone',
          sortable: 'custom',
          compare: (a, b) => a.cone - b.cone,
        },
      ],

      options: [
        {
          value: 'inputAsn',
          label: 'ASN',
        },
        {
          value: 'searchAsName',
          label: 'AS Name',
        },
        {
          value: 'searchOrg',
          label: 'Org Name',
        },
      ],
      DashboardTabs: {
        Basic: {
          label: 'Basic',
          subComponent: AsInfoTab,
          param: {},
        },
        'IPv4 Peers': {
          label: 'IPv4 Peers',
          subComponent: IpTab,
          lazy: true,
          param: {},
        },
        'IPv6 Peers': {
          label: 'IPv6 Peers',
          subComponent: IpTab,
          lazy: true,
          param: {},
        },
        'Whois': {
          label: 'Whois',
          subComponent: DashboardWhoisTab,
          lazy: true,
          param: {},
        },
      },
      searchRowHeight: 0,
      showSticktyRow: false,
    }
  },
  props: {},
  created() {
    if (Object.keys(this.$route.query).includes('asn')) {
      this.inputAsn = this.$route.query.asn
      this.inputText = this.$route.query.asn
    }
  },
  mounted() {
    this.searchRowHeight = document.getElementById('search-row').clientHeight + 10
    window.addEventListener('scroll', this.handleFloatSearch, true)
    this.getAsInfoByASN(this.inputAsn)
  },
  methods: {
    setDataInGraphData,

    handleFloatSearch() {
      const scrollTop = document.documentElement.scrollTop
      if (scrollTop >= this.searchRowHeight) {
        if (!this.showSticktyRow) {
          this.showSticktyRow = true
        }
      } else {
        if (this.showSticktyRow) {
          this.showSticktyRow = false
        }
      }
    },
    clickHandler(param) {
      const name = param.data.name.split('|')[0]
      if (name !== this.inputAsn) {
        this.getAsInfoByASN(name)
      }
    },

    async getAsInfoByList(asnList) {
      const data = await getAsInfoByList(asnList)
      return data
    },

    async getAsInfoByASN(inputAsn) {
      this.inputAsn = inputAsn
      const {
        peer_type,
        customer_prefixes,
        provider_prefixes,
        peer_prefixes,
        country_distribute,
      } = await getPeerTypeDistribute(this.inputAsn)
      this.lastUpdateTimestamp = new Date(parseInt(localStorage.getItem('lastUpdateTimestamp') * 1000))
        .toJSON()
        .split('T')[0]
      this.DashboardTabs['IPv4 Peers'].param = {
        inputAsn,
        clickHandler: this.clickHandler,
        distribution: {
          peer_type,
          customer_prefixes,
          provider_prefixes,
          peer_prefixes,
          country_distribute,
        },
        ipVersion: 4,
      }

      this.DashboardTabs['IPv6 Peers'].param = {
        inputAsn,
        clickHandler: this.clickHandler,
        distribution: {
          peer_type,
          customer_prefixes,
          provider_prefixes,
          peer_prefixes,
          country_distribute,
        },
        ipVersion: 6,
      }

      this.DashboardTabs['Basic'].param = {
        inputAsn: this.inputAsn,
      }

      this.DashboardTabs['Whois'].param = {
        inputAsn: this.inputAsn,
      }
      const asPeerTypeMap = {}
      customer_prefixes.forEach(item => {
        asPeerTypeMap[item.name] = 'C-P'
      })
      provider_prefixes.forEach(item => {
        asPeerTypeMap[item.name] = 'P-C'
      })
      peer_prefixes.forEach(item => {
        asPeerTypeMap[item.name] = 'P-P'
      })
    },
    async doSearch() {
      this.showLoading = true
      const result = await getAsInfo(this.inputAsn, this.searchAsName, this.searchOrg)
      this.lastUpdateTimestamp = new Date(parseInt(localStorage.getItem('lastUpdateTimestamp') * 1000))
        .toJSON()
        .split('T')[0]
      this.tabldata = result
      this.showLoading = false
    },
    clickAsn(asn) {
      this.inputText = asn
      this.chickSearch()
    },
    chickSearch() {
      const number_regex = /^\d+$/
      if (number_regex.test(this.inputText)) {
        this.inputAsn = this.inputText
        this.searchAsName = ''
        this.searchOrg = ''
      } else {
        this.searchAsName = this.inputText.toUpperCase()
        this.searchOrg = this.inputText.toUpperCase()
        this.inputAsn = ''
        this.tabldata = []
      }

      if (this.inputAsn === '') {
        this.doSearch()
      } else {
        this.getAsInfoByASN(this.inputAsn)
      }
      localStorage.setItem('DashboardASN',this.inputAsn)
    },
  },
}
</script>

<style lang="scss" scope>
.dashboard-container {
  padding: 22px;
  background-color: #f0f0f0;
  height: 100%;
  .chart {
    background-color: #fff;
  }

  // .pie {
  //   width: 24%;
  // }
  // .pie,
  .bar {
    height: 50vh;
  }
  .line {
    height: 20vh;
  }
  .graph {
    height: calc(80vh - 44px);
  }
  .i-e-graph {
    height: 90vh;
  }
}
.data-card-box {
  display: flex;
  flex-wrap: wrap;
}

.sticky-div {
  position: fixed;
  text-align: left;
  padding: 10px;
  z-index: 10;
  background-color: #fff;
  width: 500px;
  border-radius: 40px;
  box-shadow: 0px 10px 16px -10px #000;
}

.search-input {
  width: 500px;
}

.sticky-div .el-input-group__append,
.sticky-div .el-input-group__prepend {
  border-radius: 40px !important;
}
.sticky-div .el-input-group__prepend {
  border-top-right-radius: 0 !important;
  border-bottom-right-radius: 0 !important;
}

.sticky-div .el-input-group__append {
  border-top-left-radius: 0 !important;
  border-bottom-left-radius: 0 !important;
}

.show-search-bar {
  animation-name: fade-in;
  animation-duration: 0.5s;
}

@keyframes fade-in {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

.hidden-search-bar {
  animation-name: fade-out;
  animation-duration: 0.5s;
  z-index: -100;
  animation-fill-mode: forwards;
}

@keyframes fade-out {
  0% {
    opacity: 1;
  }

  100% {
    opacity: 0;
  }
}

.dashboard-table .el-row {
  margin-bottom: 0px !important;
}
.as-col {
  display: flex;
  justify-content: center;
  align-items: center;
}

.as {
  color: #fff;
  font-weight: bold;
  background-image: linear-gradient(to right, #7f5ac3, #af86d8);
  width: 80%;
  height: 35px;
  line-height: 35px;
  border-radius: 5px;
  box-shadow: 5px 4px 13px -7px #000;
  cursor: pointer;
}

.as:hover {
  background-image: linear-gradient(to right, #936ec3, #c39ad8);
}
.as:focus,
.as:active {
  box-shadow: -2px 4px 11px -7px #000 inset;
}

.clickedAs {
  color: #fff;
  font-weight: bold;
  background-image: linear-gradient(to right, #ee8b76, #e29a8b);
  width: 80%;
  height: 35px;
  line-height: 35px;
  border-radius: 5px;
  box-shadow: 5px 4px 13px -7px #000;
  cursor: pointer;
}

.row-cc {
  display: flex;
  align-items: center;
  background-color: #fff;
  border-radius: 5px;
  padding: 10px;
  // height: 60px;
}
.last-update {
  display: flex;
  justify-content: space-between;
  text-align: right;
  font-size: var(--middle-font-size);
  font-weight: bold;
}

.show-col {
  animation-name: fade-in;
  animation-duration: 0.5s;
}

@keyframes fade-in {
  0% {
    opacity: 0;
    transform: translate(0, 10px);
  }

  100% {
    opacity: 1;
  }
}

.show-input {
  animation-name: fade-in-from-left;
  animation-duration: 0.5s;
}

@keyframes fade-in-from-left {
  0% {
    opacity: 0;
    transform: translate(-10px, 0px);
  }

  100% {
    opacity: 1;
  }
}
</style>
