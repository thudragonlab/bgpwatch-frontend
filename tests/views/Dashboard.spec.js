import {
  shallowMount,
  mount
} from "@vue/test-utils";
import ElementUI from '@/lib/element.js'
import Dashboard from "@/views/Dashboard";
import "jest-canvas-mock";
import MyC from 'dragonlab-component/lib/dragonLab-Component.umd.js'
import Vue from "vue";
import AsInfoTab from '@/components/Dashboard/DashboardAsInfoTab.vue'
import IpTab from '@/components/Dashboard/DashboardIPTab.vue'
import { getPeerTypeDistribute, getAsInfoByList } from '@/api/DashBoardApi'
import * as testData from '../mockData/Organization'

Vue.use(ElementUI)
Vue.use(MyC)

window.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}))


jest.mock('@/api/routeMonitor', () => {
  // const _inner = [...testData]
  const mock = require('../mockData/Organization')
  return {
    getAsInfo: jest.fn(() => Promise.resolve(mock.default))
  }
})

jest.mock('@/api/DashBoardApi', () => {
  const mock = require('../mockData/Dashboard')
  const getPeerTypeDistributeMockData = mock.getPeerTypeDistribute
  const getAsInfoByListMockData = mock.getAsInfoByList
  return {
    getPeerTypeDistribute: jest.fn(() => Promise.resolve(getPeerTypeDistributeMockData)),
    getAsInfoByList: jest.fn(() => Promise.resolve(getAsInfoByListMockData)),

  }
})

const getElementByIdSpy = jest.spyOn(document, 'getElementById');
getElementByIdSpy.mockImplementation(() => ({
  clientHeight: 100
}));


describe('Dashboard.vue', () => {
  let wrapper
  const $route = {
    query: {},
  }
  beforeEach(() => {
    wrapper = shallowMount(Dashboard, {
      stubs: {
        transition: false
      },
      mocks: {
        $route,
      },
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });
  it('should render correct contents', () => {
    expect(wrapper.html()).toMatchSnapshot();
  })

  it('should call getAsInfoByASN when chickSearch is called', async () => {
    const inputAsn = '4538'
    wrapper.vm.getAsInfoByASN = jest.fn()
    wrapper.vm.inputAsn = inputAsn
    await wrapper.vm.chickSearch()
    expect(wrapper.vm.getAsInfoByASN).toBeCalledWith(inputAsn)
  })

  it('should update DashboardTabs when getAsInfoByASN is called', async () => {
    const inputAsn = '4538'
    const peerTypeData = {
      peer_type: ['type1', 'type2'],
      customer_prefixes: [{ name: 'name1' }],
      provider_prefixes: [{ name: 'name2' }],
      peer_prefixes: [{ name: 'name3' }],
      country_distribute: [{ name: 'country1' }],
    }
    const mockParam = {
      inputAsn,
      clickHandler: wrapper.vm.clickHandler,
      distribution: {
        peer_type: peerTypeData.peer_type,
        customer_prefixes: peerTypeData.customer_prefixes,
        provider_prefixes: peerTypeData.provider_prefixes,
        peer_prefixes: peerTypeData.peer_prefixes,
        country_distribute: peerTypeData.country_distribute,
      },
      ipVersion: 4,
    }
    const expectedBasicParam = { inputAsn }
    const expectedIp4Param = mockParam
    const expectedIp6Param = { ...mockParam, ipVersion: 6 }
    getPeerTypeDistribute.mockResolvedValueOnce(peerTypeData)
    await wrapper.vm.getAsInfoByASN(inputAsn)
    expect(wrapper.vm.DashboardTabs.Basic.param).toEqual(expectedBasicParam)
    expect(wrapper.vm.DashboardTabs['IPv4 Peers'].param).toEqual(expectedIp4Param)
    expect(wrapper.vm.DashboardTabs['IPv6 Peers'].param).toEqual(expectedIp6Param)
  })
  it('should call clickHandler when clickHandler is called', () => {
    const mockParam = { data: { name: '4538|peer1' } }
    wrapper.vm.clickHandler = jest.fn()
    wrapper.vm.clickHandler(mockParam)
    expect(wrapper.vm.clickHandler).toBeCalledWith(mockParam)
  })
  it('should call getAsInfoByList when getAsInfoByList is called', async () => {
    const asnList = ['4538', '1234']
    const mockData = [{ name: 'asn1' }, { name: 'asn2' }]
    getAsInfoByList.mockResolvedValueOnce(mockData)
    const result = await wrapper.vm.getAsInfoByList(asnList)
    expect(getAsInfoByList).toBeCalledWith(asnList)
    expect(result).toEqual(mockData)
  })
  it('should render AsInfoTab component when selectedTab is "Basic"', () => {
    wrapper.vm.selectedTab = 'Basic'
    expect(wrapper.findComponent(AsInfoTab).exists()).toBe(true)
  })
  it('should render IpTab component when selectedTab is "IPv4 Peers"', () => {
    wrapper.vm.selectedTab = 'IPv4 Peers'
    expect(wrapper.findComponent(IpTab).exists()).toBe(true)
  })
  it('should render IpTab component when selectedTab is "IPv6 Peers"', () => {
    wrapper.vm.selectedTab = 'IPv6 Peers'
    expect(wrapper.findComponent(IpTab).exists()).toBe(true)
  })

  it('doSearch',async () => {
    await wrapper.vm.doSearch()
    expect(wrapper.vm.tabldata).toEqual(testData.default)
  })

  
  it('clickAsn', () => {
    wrapper.vm.chickSearch = jest.fn()
     wrapper.vm.clickAsn('123')
    expect(wrapper.vm.inputText).toBe('123')
    expect(wrapper.vm.chickSearch).toHaveBeenCalled()
  })

  it('chickSearch', () => {
    const doSearch = jest.fn()
    const getAsInfoByASN = jest.fn()
    
    wrapper.vm.doSearch = doSearch
    wrapper.vm.getAsInfoByASN = getAsInfoByASN
    wrapper.vm.inputText = '123'
    wrapper.vm.chickSearch()
    expect(wrapper.vm.inputAsn).toBe('123')
    expect(wrapper.vm.searchAsName).toBe('')
    expect(wrapper.vm.searchOrg).toBe('')
    expect(getAsInfoByASN).toHaveBeenCalled()
    wrapper.vm.inputText = 'cernet'
    wrapper.vm.chickSearch()
    expect(wrapper.vm.inputAsn).toBe('')
    expect(wrapper.vm.searchAsName).toBe('CERNET')
    expect(wrapper.vm.searchOrg).toBe('CERNET')
    expect(doSearch).toHaveBeenCalled()
    


  })
  describe('handleFloatSearch method', () => {
    it('should set showSticktyRow to true when scrollTop is greater than or equal to searchRowHeight', () => {
      wrapper.vm.showSticktyRow = false
      wrapper.vm.searchRowHeight = 100
      document.documentElement.scrollTop = 150
      wrapper.vm.handleFloatSearch({ target: { scrollTop: 150 } })
      expect(wrapper.vm.showSticktyRow).toBe(true)
    })
    it('should set showSticktyRow to false when scrollTop is less than searchRowHeight', () => {
      wrapper.vm.showSticktyRow = true
      wrapper.vm.searchRowHeight = 100
      document.documentElement.scrollTop = 50
      wrapper.vm.handleFloatSearch({ target: { scrollTop: 50 } })
      expect(wrapper.vm.showSticktyRow).toBe(false)
    })
  })
  describe('clickHandler method', () => {
    it('should set inputAsn correctly when name is not equal to inputAsn', () => {
      const paramMock = { data: { name: 'test|123' } }
      wrapper.vm.inputAsn = 'abc'
      wrapper.vm.getAsInfoByASN = jest.fn()
      wrapper.vm.clickHandler(paramMock)
      expect(wrapper.vm.inputAsn).toBe('abc')
      expect(wrapper.vm.getAsInfoByASN).toHaveBeenCalledWith('test')
    })
    it('should not call getAsInfoByASN method when name is equal to inputAsn', () => {
      const paramMock = { data: { name: 'abc|123' } }
      wrapper.vm.inputAsn = 'abc'
      wrapper.vm.getAsInfoByASN = jest.fn()
      wrapper.vm.clickHandler(paramMock)
      expect(wrapper.vm.getAsInfoByASN).not.toHaveBeenCalled()
    })

  })
})

describe('create Dashboard',() => {
  const $route = {
    query: {
      asn:'174'
    },
  }
  it('create',() => {
    const wrapper = shallowMount(Dashboard, {
      stubs: {
        transition: false
      },
      mocks: {
        $route,
      },
    });
    expect(wrapper.vm.inputAsn).toBe('174')
    expect(wrapper.vm.inputText).toBe('174')
  })
})