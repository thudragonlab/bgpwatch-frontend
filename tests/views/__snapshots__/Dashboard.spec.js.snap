// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard.vue should render correct contents 1`] = `
<div class="dashboard-container fade-in">
  <div class="sticky-div hidden-search-bar">
    <el-input-stub value="4538" type="text" validateevent="true" placeholder="You can search by AS number, AS name, organization name" class="search-input">
      <el-button-stub type="default" icon="el-icon-search" nativetype="button"></el-button-stub>
    </el-input-stub>
  </div>
  <el-row-stub tag="div" justify="start" id="search-row" class="row-cc">
    <el-col-stub span="9" tag="div" class="show-input" style="text-align: left;">
      <el-input-stub value="4538" type="text" validateevent="true" placeholder="You can search by AS number, AS name, organization name" class="search-input">
        <el-button-stub type="default" icon="el-icon-search" nativetype="button"></el-button-stub>
      </el-input-stub>
    </el-col-stub>
    <el-col-stub span="15" tag="div" class="last-update">
      <div style="text-align: left;">You can search by AS number, AS name, or organization name.</div>
      <div>Last Update:1970-01-01</div>
    </el-col-stub>
  </el-row-stub>
  <!---->
  <el-row-stub tag="div" justify="start">
    <el-tabs-stub type="border-card" value="Basic" tabposition="top" stretch="true" class="tab-list">
      <el-tab-pane-stub label="Basic" name="Basic">
        <asinfotab-stub param="[object Object]"></asinfotab-stub>
      </el-tab-pane-stub>
      <el-tab-pane-stub label="IPv4 Peers" name="IPv4 Peers" lazy="true">
        <iptab-stub param="[object Object]"></iptab-stub>
      </el-tab-pane-stub>
      <el-tab-pane-stub label="IPv6 Peers" name="IPv6 Peers" lazy="true">
        <iptab-stub param="[object Object]"></iptab-stub>
      </el-tab-pane-stub>
      <el-tab-pane-stub label="Whois" name="Whois" lazy="true">
        <dashboardwhoistab-stub param="[object Object]"></dashboardwhoistab-stub>
      </el-tab-pane-stub>
    </el-tabs-stub>
  </el-row-stub>
</div>
`;
